### PR TITLE
Outfit details handles negative values as "needed" better

### PR DESCRIPTION
**Pull request by alextd**
## Feature Details
The outfit display now uses a general rule: negative attributes are requirements, and are shown at the end of the list (Some exceptions exist)

This ends up looking mostly the same as before, as the most common negative values were already handled as an exception: "outfit space", "weapon capacity" etc. They were already specially chosen to go at the end of the list, and tagged with the "needed".

So now, it's just a general rule: any negative value is considered a requirement and shown there in the same way alongside "outfit space needed" - e.g.  "missile capacity needed: 1" instead of "missile capacity: -1" at the top. So the rare time you have an outfit that uses cargo space, it'll show "cargo space needed: 20" instead of "cargo space: -20"

*edit: There are some exceptions (to be discussed)*
~The global exceptions are "automaton" and "heat dissipation". They're just shown as-is. The other exception is "crew needed: 1" which is shown alongside other requirements.~

Any attribute that is usually a requirement, but is positive for this outfit, is shown like "outfit space added: 15" instead of "outfit space needed: -15" in the requirement area.

## UI Screenshots
The Outfits Expansion gets the most benefit:
imageimage

Turrets with crew:
imageimage

The special requirement for the Dragonflame Cannon:
imageimage

Ammunition: 
imageimage

'automaton' is not affected:
imageimage

## Usage Examples
N/A, it just looks nicer

## Testing Done
Looked at a lot of outfits! Easy peasy!

## Performance Impact
N/A

### CODE REVIEW

**Comment by alextd** [comment]
I only just now thought, "what does the map screen look like"

image
image

So guess that is affected, but it doesn't show the requirements part. Huh.

**Comment by EjoThims** [comment]
I like this, even the map panel difference.

**Comment by alextd** [comment]
It's really just because the blanket rule is to append `" added:"` for every other attribute, so displaying `required crew: 1` would be yet another exception to add code for.

I'll put up a mockup of requirements at the beginning. Right now, code bundles those lines with cost, so that'll take a small chunk of work.

**Comment by alextd** [comment]
Possible different looks. 

Honestly, cost at the top is feeling fine. Which... is how it already is on the Ship screen. Huh.
image
image
image

**Comment by alextd** [comment]
So that was CRA. Cost, Requirements, Attributes. ACR is how it already is.

How about RAC?
image

ARC?
image

PS: That last one was photoshopped. For some reason I've been coding up actual changes instead of just copy-pasting some rectangles.

**Comment by Amazinite** [comment]
I prefer CRA. I've actually had switching it up to that order on my to-do list.

File: source/OutfitInfoDisplay.cpp [hunk 1]
```diff
@@ -272,33 +279,59 @@ void OutfitInfoDisplay::UpdateRequirements(const Outfit &outfit, const PlayerInf
...
+// Any exceptions to that rule would require in-game code to handle
+// their unique properties, so when code is added to handle a new
+// attribute, this code also should also be updated.
+void OutfitInfoDisplay::AddRequirementAttribute(string label, double value)
+{
+	// These attributes have negative values but are not requirements
+	if(IsNotRequirement(label))
+		return;
+
+	// Special case for 'crew required' - use positive values as a requirement.
```
Comment (Hurleveur) [inline source/OutfitInfoDisplay.cpp:322]:
```suggestion
	// Special case for 'required crew' - use positive values as a requirement.
```

File: source/OutfitInfoDisplay.cpp [hunk 2]
```diff
@@ -192,6 +192,13 @@ namespace {
 		{"minable", "This item is mined from asteroids."},
 		{"atrocity", "This outfit is considered an atrocity."}
 	};
+
+	bool IsNotRequirement(const string& label)
+	{
+		return label == "automaton" ||
+			SCALE.find(label) != SCALE.end() ||
+			BOOLEAN_ATTRIBUTES.find(label) != BOOLEAN_ATTRIBUTES.end();
+	}
```
Comment (Hurleveur) [inline source/OutfitInfoDisplay.cpp:201]: Looking at the scales and booleans seems like a weird way of doing it? I'm not saying it is necessarily wrong, I'm just not sure I understand why it is the case. We could instead use a list of actual requirements?

I see line 348-349 saying `They conveniently don't use SCALE or BOOLEAN_ATTRIBUTES.` which doesn't seem very resilient as a way of handling it.
This function could at least use a good comment in front to explain why it works this way, which idk if it should.


Also, automaton is a boolean ship attribute; so you don't have to consider it as an exception; if at some point it gets added to outfits it'll be a part of `BOOLEAN_ATTRIBUTES`.

**Review by Hurleveur — CHANGES_REQUESTED** [review]
My previous comments were addressed.

Looking at the code the way an attribute is determined to be or not a requirement puzzles me, maybe there are actual reasons but then those should be clear in the form of comments so that it's easy to modify this code in the future.
If it is just a coincidence that nothing in there is a requirement that is probably not the way to go.

File: source/OutfitInfoDisplay.cpp (hunk 2 above)
Comment (alextd) [reply source/OutfitInfoDisplay.cpp:201]: So I'm looking at this, and re-writing this as "IsRequirement" which also takes the value. This started as a little exception but the end result here is not pretty. So a few parts are getting consolidated in here. And the comment will go here to show why it's listed as a requirement.

> They conveniently don't use SCALE or BOOLEAN_ATTRIBUTES

That was supposed to mean "doesn't need formatting so the code is to display is easy". It's not really needed to mention that. Nevermind.

File: source/OutfitInfoDisplay.cpp [hunk 3]
```diff
@@ -310,12 +343,46 @@ void OutfitInfoDisplay::UpdateAttributes(const Outfit &outfit)
...
+			if(it.second > 0)
+				continue;
+
+			// A negative 'required crew' would be a benefit, so it is listed here.
+		}
+		else if(IsNotRequirement(it.first))
+		{
+			// This attribute is always listed here, though they may be negative.
+		}
+		else if(it.second < 0)
```
Comment (RisingLeaf) [inline source/OutfitInfoDisplay.cpp:385]:
```suggestion
		else if(it.second < 0 && !IsNotRequirement(it.first))
```
This is just better than having an empty if.

**Review by RisingLeaf — APPROVED** [review]
Looks nice overall

File: source/OutfitInfoDisplay.cpp [hunk 4]
```diff
@@ -206,6 +206,13 @@ namespace {
 		{"minable", "This item is mined from asteroids."},
 		{"atrocity", "This outfit is considered an atrocity."}
 	};
+
+	bool IsNotRequirement(const string& label)
```
Comment (warp-core) [inline source/OutfitInfoDisplay.cpp:210]:
```suggestion
	bool IsNotRequirement(const string &label)
```

File: source/OutfitInfoDisplay.cpp [hunk 5]
```diff
@@ -324,12 +357,46 @@ void OutfitInfoDisplay::UpdateAttributes(const Outfit &outfit)
...
 	bool hasNormalAttributes = false;
+
+	// These attributes are regularly negative on outfits, so when positive,
+	// tag them with "added" and show them first. They conveniently
+	// don't use SCALE or BOOLEAN_ATTRIBUTES.
+	static const vector<string> EXPECTED_NEGATIVE = {
+		"outfit space", "weapon capacity", "engine capacity", "gun ports", "turret mounts"
+	};
+
+	for(const string& attr : EXPECTED_NEGATIVE)
```
Comment (warp-core) [inline source/OutfitInfoDisplay.cpp:368]:
```suggestion
	for(const string &attr : EXPECTED_NEGATIVE)
```

**Review by warp-core — COMMENTED** [review]
Style fixes

**Review by warp-core — APPROVED** [review]

**Pull request merged**